src/common/argsregister.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_COMMON_ARGSREGISTER_H
+#define BITCOIN_COMMON_ARGSREGISTER_H
+
+#include <common/args.h>
+
+#include <type_traits>
+
+/** Typesafe args registration.  */
+
+/** Example usage:
+
+constexpr bool DEFAULT_FOO_A{true};
+
+struct FooOpts {
+    bool a{DEFAULT_FOO_A};
+    std::optional<std::string> b;
+    std::vector<std::string> c;
+    Custom d;
+};
+
+class FooRegister
+{
+public:
+    using T = CChainParams::SigNetOptions;
+
+    static inline void GetD(Custom& d, const std::string& arg_d);
+
+    template<typename C, typename Op>
+    static inline void Register(Op& op)
+    {
+        return C::Do(op,
+            C::Defn(&T::a, "-fooa", "",
+                    "Description of option A",
+                    ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION,
+                    OptionsCategory::DEBUG_TEST),
+            C::Defn(&T::b, "-foob", "=bar",
+                    "Description of option B",
+                    ArgsManager::ALLOW_ANY,
+                    OptionsCategory::DEBUG_TEST),
+            C::Defn(&T::c, "-foob", "=bar",
+                    "Description of option B",
+                    ArgsManager::ALLOW_ANY,
+                    OptionsCategory::DEBUG_TEST),
+            C::Defn(&T::d, "-foob", "=bar", GetD,
+                    "Description of option B",
+                    ArgsManager::ALLOW_ANY,
+                    OptionsCategory::DEBUG_TEST)
+            );
+    }
+};
+
+void SetupFooOptions(ArgsManager& argsman)
+{
+    ArgsRegister<FooArgsRegister>::Register(argsman);
+}
+
+void ReadFooArgs(const ArgsManager& args, FooOpts& options)
+{
+    ArgsRegister<FooArgsRegister>::Update(args, options);
+}
+
+**/
+
+template<typename REG>
+class ArgsRegister
+{
+public:
+    using STRUCT = typename REG::T;
```

Comment:
(Using `ArgsRegisterImpl` naming would make this be `template<typename STRUCT> class ArgsRegister { public: using REG = ArgsRegisterImpl<T>;` instead)